hail/python/hail/ggplot/geoms.py

```diff
@@ -166,24 +170,43 @@ def _add_legend(self, fig_so_far: go.Figure, aes_name, category, value):
             }
         )
 
-    def apply_to_fig(self, grouped_data, fig_so_far: go.Figure, precomputed, facet_row, facet_col, legend_cache) -> bool:
+    def apply_to_fig(self, grouped_data, fig_so_far: go.Figure, precomputed, facet_row, facet_col, legend_cache, is_faceted) -> bool:
+        traces = []
         legends = {}
         for df in grouped_data:
             values = self._get_aes_values(df)
-            self._add_trace(fig_so_far, df, facet_row, facet_col, values)
+            trace_category = None
             for aes_name in self.aes_legend_groups:
+                category = self._get_aes_value(df, f"{aes_name}_legend")
+                trace_category = category if category is not None else trace_category
                 legends[aes_name] = ({
                     **legends.get(aes_name, {}),
-                    self._get_aes_value(df, f"{aes_name}_legend"): values[aes_name]
+                    category: values[aes_name]
                 })
-
-        number_of_displayed_legends = 0
-        for aes_name, legend_group in legends.items():
-            if len(legend_group) > 1:
-                number_of_displayed_legends += 1
+            traces = [*traces, [fig_so_far, df, facet_row, facet_col, values, trace_category]]
+
+        if is_faceted:
+            requires_static = True
+        else:
+            count = 0
+            for legend_group in legends.values():
+                count += 1 if len(legend_group) > 1 else 0
+            requires_static = count > 1
+
+        if requires_static:
+            for trace in traces:
+                self._add_trace(*trace[:-1])
```

Comment:
This is a bit too clever 😉 . What's the harm in sending the trace_category into `_add_trace` in the static case?